common/persistence/sql/nexus_service_store.go

```diff
@@ -114,44 +114,53 @@ func (s *sqlNexusIncomingServiceStore) ListNexusIncomingServices(
 	}
 
-	resp, err := s.Db.ListNexusIncomingServices(ctx, &sqlplugin.ListNexusIncomingServicesRequest{
-		LastKnownTableVersion: request.LastKnownTableVersion,
-		LastServiceID:         lastServiceID,
-		Limit:                 request.PageSize,
+	var response p.InternalListNexusIncomingServicesResponse
+	var rows []sqlplugin.NexusIncomingServicesRow
+	retErr := s.txExecute(ctx, "ListNexusIncomingServices", func(tx sqlplugin.Tx) error {
+		curTableVersion, err := tx.GetNexusIncomingServicesTableVersion(ctx)
+		if err != nil {
+			return err
+		}
+		response.TableVersion = curTableVersion
+		if request.LastKnownTableVersion != 0 && request.LastKnownTableVersion != curTableVersion {
+			return p.ErrNexusTableVersionConflict
+		}
+
+		rows, err = tx.ListNexusIncomingServices(ctx, &sqlplugin.ListNexusIncomingServicesRequest{
+			LastServiceID: lastServiceID,
+			Limit:         request.PageSize,
+		})
+
+		return err
 	})
-	if err != nil {
-		if errors.Is(err, p.ErrNexusIncomingServiceVersionConflict) {
-			// On table version conflict, return current table version and appropriate error
-			return &p.InternalListNexusIncomingServicesResponse{TableVersion: resp.CurrentTableVersion}, err
+
+	if retErr != nil {
+		if errors.Is(retErr, p.ErrNexusTableVersionConflict) {
+			return &response, retErr
 		}
-		// For all other errors, operation failed so return Unavailable error
-		s.logger.Error("ListNexusIncomingServices operation failed", tag.Error(err))
-		return nil, serviceerror.NewUnavailable(fmt.Sprintf("ListNexusIncmoingServices operation failed: %v", err))
+		return nil, retErr
 	}
 
 	var nextPageToken []byte
-	if len(resp.Entries) == request.PageSize {
-		nextPageToken, err = serializePageTokenJson(&listIncomingServicesNextPageToken{
-			LastServiceID: resp.Entries[request.PageSize-1].ServiceID,
+	if len(rows) == request.PageSize {
+		nextPageToken, retErr = serializePageTokenJson(&listIncomingServicesNextPageToken{
+			LastServiceID: rows[request.PageSize-1].ServiceID,
 		})
-		if err != nil {
-			s.logger.Error("error serializing next page token during ListNexusIncomingServices", tag.Error(err))
-			return nil, serviceerror.NewInternal(err.Error())
+		if retErr != nil {
+			s.logger.Error("error serializing next page token during ListNexusIncomingServices", tag.Error(retErr))
+			return nil, serviceerror.NewInternal(retErr.Error())
 		}
 	}
+	response.NextPageToken = nextPageToken
 
-	services := make([]p.InternalNexusIncomingService, len(resp.Entries))
-	for i, entry := range resp.Entries {
-		services[i].ServiceID = entry.ServiceID
-		services[i].Version = entry.Version
-		services[i].Data = p.NewDataBlob(entry.Data, entry.DataEncoding)
+	response.Services = make([]p.InternalNexusIncomingService, len(rows))
+	for i, row := range rows {
+		response.Services[i].ServiceID = primitives.UUIDString(row.ServiceID)
+		response.Services[i].Version = row.Version
+		response.Services[i].Data = p.NewDataBlob(row.Data, row.DataEncoding)
```

Comment:
nit: you can use the row as a field in `InternalNexusIncomingService` if that works with the cass implementation.

Reply:
I think I am going to make some changes to `InternalNexusIncomingService` when implementing the manager-level, so I will leave it as is for now.